### DISCUSSION TITLE
'Update' on JSON field preserves the existing JSON data

### DISCUSSION THREAD

**Discussion by Redzli** [prisma/prisma#3070] [Q&A] [answered]
Hey, when I run an `update` statement on a table that has a JSON field, I would like to `update` the JSON field that has some existing data, instead of "replacing" it. What I mean by this is to append another key-value pair to the JSON field, while preserving the old data in the JSON.

My schema: 
```prisma
model User {
  userMetadata   Json?
```

When I created the user, I do
```javascript
const createUser = await prisma.user.create({
  data: {
    userMetadata: {
       name: "John"
    }
  }
});
```

Then when I want to update his metadata, I do:

```javascript
  await prisma.user.update({
    where: {
      id: userId //suppose there is one
    },
    data:{
      userMetadata: {
        university: "Oxford"
      }
    }
  })
  ```

## Expected results
Ideally after running the `update` I would like to see the user having such `userMetadata`:
```
userMetadata: {
  name: "John",
  university: "Oxford"
}
```

## Actual results
However, it only shows 

```
userMetadata: {
  university: "Oxford"
}
```

## Question
I am wondering if there is a way to append new key-value pairs like `university: "Oxford"` to the userMetadata while preserving the old data?

**Comment by ryands17** [comment, marked as answer]
Heu @Redzli 👋 
Unfortunately direct update on JSON fields isn't supported yet, as a workaround, you would need to fetch the json first, append to it for any changes and then call the updated method. I think you have created a request for that here as well.
https://github.com/prisma/prisma-client-js/issues/790

**Reply by Redzli** [reply]
Yep, thanks for the answer

**Reply by jasondemorrow** [reply]
This by itself isn't an acceptable workaround since, even if you were to fetch the "current" data, this could have itself been updated by another request in between your retrieval and update, meaning that you will still overwrite data in other fields.

**Comment by snnkzn** [comment]
Hi try this;

data: { userMetadata: { ...userMetadata, university:  "Oxford"} },

**Reply by piradata** [reply]
why the downvotes? it actually works

```js
for (const application of applicationsToUpdate) {
        await this.prisma.application.update({
          where: {
            id: application.id,
          },
          data: {
            data: {
              ...application.data,
              targetInstallationCode: newTargetInstalationCode
            },
         },
    });
}
```

**Reply by ScreamZ** [reply]
Because this requires to fetch the actual record, before updating it.

**Reply by jasondemorrow** [reply]
Ironically, this massively downvoted answer is practically identical to the accepted answer, if you assume `userMetadata` was retrieved just prior. It suffers the same fundamental flaw in that it will not reliably solve the problem in the face of multiple, concurrent requests.

**Comment by krismartin** [comment]
A bit too late to bring this up now, but have you considered using [jsonb_set](https://www.postgresql.org/docs/current/functions-json.html) (assuming you're using Postgres)?

Something like this:
```javascript
const sql = Prisma.sql`
  update "User"
  set userMetadata = jsonb_set(userMetadata, '{university}', '"${Prisma.raw(universitryName)}"', true)
  where id = ${userId};
  `

await prisma.$queryRaw(sql)
```

**Reply by gschechter** [reply]
Yes this works but you can do this with any SQL database provider,  it abandons the purpose of the ORM.
